irrd/server/tests/test_query_resolver.py

```diff
@@ -554,6 +564,58 @@ def test_route_set_compatibility_ipv4_only_route_set_members(self, prepare_resol
         result = resolver.members_for_set('RS-TEST', recursive=False)
         assert result == ['192.0.2.0/32', '192.0.2.1/32', 'RS-OTHER']
 
+    def test_members_for_set_per_source(self, prepare_resolver):
```

Comment:
This test should be expanded specifically to verify deeper resolving, because it is very important that the limitation on the root source is not included on later expansion queries.